File: core/contractsdk/java/example/charity_demo/src/main/java/com/baidu/xuper/example/CharityDemo.java
package com.baidu.xuper.example;

import java.math.BigDecimal;
import java.util.concurrent.atomic.AtomicInteger;

import com.baidu.xuper.Context;
import com.baidu.xuper.Contract;
import com.baidu.xuper.ContractMethod;
import com.baidu.xuper.Driver;
import com.baidu.xuper.Response;
import com.baidu.xuper.utils.ByteUtils;
import com.sun.tools.javac.util.ByteBuffer;

/**
 * Counter
 */
public class CharityDemo implements Contract {
    final String USERDONATE = "UserDonate_";
    final String ALLDONATE = "AllDonate_";
    final String ALLCOST = "AllCost_";
    final String TOTALRECEIVED = "TotalDonates";
    final String TOTALCOSTS = "TotalCosts";
    final String BALANCE = "Balance_";
    final String DONATECOUNT = "DonateCount_";
    final String COSTCOUNT = "CostCount_";
    final String ADMIN = "admin";
    final int MAX_LIMIT = 100; // uint64???


    @Override
    @ContractMethod
    public Response initialize(Context ctx) {
        byte[] admin = ctx.args().get("admin".getBytes());
        ctx.putObject(ADMIN.getBytes(), admin);
        ctx.putObject(TOTALRECEIVED.getBytes(), "0".getBytes());
        ctx.putObject(TOTALCOSTS.getBytes(), "0".getBytes());
        ctx.putObject(BALANCE.getBytes(), "0".getBytes());
        ctx.putObject(DONATECOUNT.getBytes(), "0".getBytes());
        ctx.putObject(COSTCOUNT.getBytes(), "0".getBytes());
        return Response.ok("ok".getBytes());
    }

    private boolean isAdmin(Context ctx) {
        String caller = ctx.caller();
        return new String(ctx.getObject(ADMIN.getBytes())).equals(caller);
    }

    private String getIdFromNum(BigDecimal num) {

        String numStr = num.toString();
        if (numStr.length() >= 20) {
            return numStr;
        }
        int count = 20 - numStr.length(); // magic number is from cpp examples
        return new String(new char[count]).replace('\0', '0') + numStr;
    }

    @ContractMethod
    public Response donate(Context ctx) {
        if (!this.isAdmin(ctx)) {
            return Response.error("only admin can add donate record");
        }
        String donor = new String(ctx.args().get("donor"));
        BigDecimal amount = new BigDecimal(new String(ctx.args().get("amount")));
        String timestamp = new String(ctx.args().get("timestamp"));
        String comments = new String(ctx.args().get("comments"));
        if (donor.length() == 0 || timestamp.length() == 0 || comments.length() == 0) {
            return Response.error("missing donor or amount or timestamp or comments");
        }
        if (amount.compareTo(BigDecimal.ZERO) <= 0) {
            return Response.error("amount must be positive");
        }
        BigDecimal donateCount = new BigDecimal(new String(ctx.getObject(DONATECOUNT.getBytes())));
        BigDecimal totalRecv = new BigDecimal(new String(ctx.getObject(TOTALRECEIVED.getBytes())));
        BigDecimal balance = new BigDecimal(new String(ctx.getObject(BALANCE.getBytes())));
        totalRecv = totalRecv.add(amount);
        balance = balance.add(amount);
        donateCount = donateCount.add(BigDecimal.ONE);

        String donateId = getIdFromNum(donateCount);

        String userDonateKey = USERDONATE + donor + "%" + donateId;
        String allDonateKey = (ALLDONATE + donateId);
        String donateDetail = "donor=" + donor +
                ",amount=" + amount +
                "timestamp=" + timestamp +
                "comments = " + comments;
        ctx.putObject(userDonateKey.getBytes(), donateDetail.getBytes());
        ctx.putObject(allDonateKey.getBytes(), donateDetail.getBytes());
        ctx.putObject(DONATECOUNT.getBytes(), donateCount.toString().getBytes());
        ctx.putObject(TOTALRECEIVED.getBytes(), totalRecv.toString().getBytes());
        ctx.putObject(BALANCE.getBytes(), balance.toString().getBytes());
        return Response.ok(donateId.getBytes());
    }

    @ContractMethod
    public Response cost(Context ctx) {
        if (!this.isAdmin(ctx)) {
            return Response.error("only admin can add cost record");
        }
        String to = new String(ctx.args().get("to"));
        BigDecimal amount = new BigDecimal(new String(ctx.args().get("amount")));
        String timestamp = new String(ctx.args().get("timestamp"));
        String comments = new String(ctx.args().get("comments"));

        if (to.length() == 0 || timestamp.length() == 0 || comments.length() == 0) {
            return Response.error("missing to or amount or timestamp or comments");
        }
        if (amount.compareTo(BigDecimal.ZERO) <= 0) {
            return Response.error("amount must be positive");
        }

        String totalCost = new String(ctx.getObject(TOTALCOSTS.getBytes()));
        BigDecimal balance = new BigDecimal(new String(ctx.getObject(BALANCE.getBytes())));
        BigDecimal costCount = new BigDecimal(new String(ctx.getObject(COSTCOUNT.getBytes())));
        if (balance.compareTo(amount) < 0) {
            return Response.error("fund balance not enough");
        }


        String costId = getIdFromNum(costCount);
        String allCostKey = ALLCOST + costId;
        String costDetail = "to=" + to +
                ",amount=" + amount +
                ",timestamp=" + timestamp +
                ",comments=" + comments;


        ctx.putObject(allCostKey.getBytes(), costDetail.getBytes());
        ctx.putObject(COSTCOUNT.getBytes(), costCount.toString().getBytes()); // TODO 好像差一个
        ctx.putObject(TOTALCOSTS.getBytes(), totalCost.getBytes());
        ctx.putObject(BALANCE.getBytes(), balance.toString().getBytes());
        return Response.ok(costId.getBytes());
    }

    @ContractMethod
    public Response statistics(Context ctx) {
        String totalCost = new String(ctx.getObject(TOTALCOSTS.getBytes()));
        String balance = new String(ctx.getObject(BALANCE.getBytes()));
        String totalRec = new String(ctx.getObject(TOTALRECEIVED.getBytes()));
        String result =
                "totalDonates=" + totalRec + "," +
                        "totalCosts=" + totalCost + "," +
                        "fundBalance=" + balance;
        return Response.ok(result.getBytes());
    }

    @ContractMethod
    public Response queryDonor(Context ctx) {
        String donor = new String(ctx.args().get("donor".getBytes()));
        if (donor.length() == 0) {
            return Response.error("missing donor argument");
        }
        String userDonateKey = USERDONATE + donor + "%";
        ByteBuffer buf = new ByteBuffer();
        AtomicInteger donateCount = new AtomicInteger(); // TODO @fengjin
        ctx.newIterator(userDonateKey.getBytes(), (userDonateKey + "~").getBytes()).forEachRemaining(
                elem -> {
//                    TODO 验证
                    donateCount.getAndIncrement();
                    String donateId = new String(elem.getKey()).substring(userDonateKey.length());
                    String content = ByteUtils.toString(elem.getValue());
                    buf.appendBytes((
                            "id=" + donateId + "," +
                                    "content=" + content + "\n"
                    ).getBytes());
                }
        );
        return Response.ok(buf.elems); // TODO @fengjin

    }

    @ContractMethod
    public Response queryDonates(Context ctx) {
        String  startId = new String(ctx.args().get("startid"));
        BigDecimal  limit = new BigDecimal( new String(ctx.args().get("limit")));
        if (startId.length() == 0) {
            return Response.error("missing startid or limit");
        }
        if (limit.compareTo(BigDecimal.ZERO) <=0){
            return Response.error("limit must be positive");
        }

        if (limit.compareTo( new BigDecimal(MAX_LIMIT))>=0) {
            return Response.error("limit must be less than" + MAX_LIMIT);
        }

        String  donateKey = ALLDONATE+ startId;
        int limit1 = 0;
        AtomicInteger selected = new AtomicInteger();
        ByteBuffer buf = new ByteBuffer();
        ctx.newIterator(donateKey.getBytes(), (donateKey+ "~").getBytes()).forEachRemaining(
                elem -> {
                    if (selected.get() >= limit1) {// 注意边界条件
                        return;
                    }
//                    TODO 检查
                    selected.getAndIncrement();
                    String donateId = ByteUtils.toString(elem.getKey()).substring(ALLDONATE.length());
                    String content = ByteUtils.toString(elem.getValue());
                    buf.appendBytes((
                            "id=" + donateId + "," +
                                    content + "\n" // TODO
                    ).getBytes());
                }
        );
        return Response.ok(buf.elems);
    }

    @ContractMethod
    public Response queryCosts(Context ctx) {
        String  startId = new String(ctx.args().get("startid"));
        BigDecimal  limit = new BigDecimal( new String(ctx.args().get("limit")));
        if (startId.length() == 0) {
            return Response.error("missing startid or limit");
        }
        if (limit.compareTo(BigDecimal.ZERO) <=0){
            return Response.error("limit must be positive");
        }

        if (limit.compareTo( new BigDecimal(MAX_LIMIT))>=0) {
            return Response.error("limit must be less than" + MAX_LIMIT);
        }

        String donateKey = ALLCOST+startId;
        int limit1 = 0; // TODO
        AtomicInteger selected = new AtomicInteger();
        ByteBuffer buf = new ByteBuffer();
        ctx.newIterator(donateKey.getBytes(), (donateKey+ "~").getBytes()).forEachRemaining(
                elem -> {
                    if (selected.get() >= limit1) {// 注意边界条件
                        return;
                    }
//                    TODO 检查
                    selected.getAndIncrement();
                    String donateId = ByteUtils.toString(elem.getKey()).substring(ALLCOST.length());
                    String content = ByteUtils.toString(elem.getValue());
                    buf.appendBytes((
                            "id=" + donateId + "," +
                                    content + "\n" // TODO
                    ).getBytes());
                }
        );
        return Response.ok(buf.elems);
    }


    public static void main(String[] args) {
        Driver.serve(new CharityDemo());
    }
}

File: core/contractsdk/cpp/example/charity_demo.cc
#include <iomanip>
#include <sstream>
#include "xchain/xchain.h"

// 慈善捐款公示模板
// 参数由xchain::Contract中的context提供
class Charity {
public:
    /*
     * func: 初始化慈善基金管理账户
     * @param: admin: 哪个address具有管理员权限
     */
    virtual void initialize() = 0;
    /*
     * func: 新增捐款
     * @param: donor:捐款人id（注意不能包含/字符）
     * @param: amount: 捐款金额
     * @param: timestamp: 捐款时间
     * @param: comments: 备注
     * @return: donateid: 捐款编号
     */
    virtual void donate() = 0;
    /*
     * func: 新增慈善花费
     * @param: to:善款受益人
     * @param: amount: 善款金额
     * @param: timestamp: 拨款时间
     * @param: comments: 备注，例如受益人接收证明(可以是收据链接)
     * @return: costid: 拨款编号
     */
    virtual void cost() = 0;
    /*
     * func: 获取善款综述
     * @return: totalDonates(总捐款金额),
     * totalCosts(总拨付善款),fundBalance(基金会善款余额)
     */
    virtual void statistics() = 0;
    /*
     * func: 查询某个用户的捐款记录
     * @param: donor:捐款人id
     */
    virtual void queryDonor() = 0;
    /*
     * func: 查询捐款记录
     * @param: startid: 起始记录id
     * @param: limit: 查询多少条(每次查询不超过100条)
     */
    virtual void queryDonates() = 0;
    /*
     * func: 查询拨款记录
     * @param: startid: 起始记录id
     * @param: limit: 查询多少条(每次查询不超过100条)
     */
    virtual void queryCosts() = 0;
};

struct CharityDemo : public Charity, public xchain::Contract {
private:
    const std::string USERDONATE = "UserDonate_";
    const std::string ALLDONATE = "AllDonate_";
    const std::string ALLCOST = "AllCost_";
    const std::string TOTALRECEIVED = "TotalDonates";
    const std::string TOTALCOSTS = "TotalCosts";
    const std::string BALANCE = "Balance";
    const std::string DONATECOUNT = "DonateCount";
    const std::string COSTCOUNT = "CostCount";
    const std::string ADMIN = "admin";
    const uint64_t MAX_LIMIT = 100;

    std::string getIDFromNum(uint64_t num) {
        std::ostringstream ss;
        ss << std::setw(20) << std::setfill('0') << num;
        return ss.str();
    }

    bool safe_stoull(const std::string in, uint64_t* out) {
        if (in.empty()) {
            return false;
        }
        for (int i = 0; i < in.size(); i++) {
            if (in[i] < '0' || in[i] > '9') {
                return false;
            }
        }
        std::string::size_type sz = 0;
        *out = std::stoull(in, &sz);
        if (sz != in.size()) {
            return false;
        }
        return true;
    }

    bool isAdmin(xchain::Context* ctx, const std::string& caller) {
        std::string admin;
        if (!ctx->get_object(ADMIN, &admin)) {
            return false;
        }
        return (admin == caller);
    }

public:
    void initialize() {
        xchain::Context* ctx = this->context();
        const std::string& admin = ctx->arg(ADMIN);
        if (admin.empty()) {
            ctx->error("missing admin address");
            return;
        }
        ctx->put_object(ADMIN, admin);
        ctx->put_object(TOTALRECEIVED, "0");
        ctx->put_object(TOTALCOSTS, "0");
        ctx->put_object(BALANCE, "0");
        ctx->put_object(DONATECOUNT, "0");
        ctx->put_object(COSTCOUNT, "0");
        ctx->ok("initialize success");
    }

    void donate() {
        xchain::Context* ctx = this->context();
        const std::string& caller = ctx->initiator();
        if (caller.empty()) {
            ctx->error("missing initiator");
            return;
        }

        if (!isAdmin(ctx, caller)) {
            ctx->error("only the admin can add donate record");
            return;
        }

        const std::string& donor = ctx->arg("donor");
        if (donor.empty()) {
            ctx->error("missing 'donor'");
            return;
        }

        const std::string& amountStr = ctx->arg("amount");
        if (amountStr.empty()) {
            ctx->error("missing 'amount'");
            return;
        }

        uint64_t amount;
        if (!safe_stoull(amountStr, &amount)) {
            ctx->error(
                "illegel 'amount', should be string of a positive number");
            return;
        }

        const std::string& timestamp = ctx->arg("timestamp");
        if (timestamp.empty()) {
            ctx->error("missing 'timestamp'");
            return;
        }

        const std::string& comments = ctx->arg("comments");
        if (comments.empty()) {
            ctx->error("missing 'comments'");
            return;
        }

        uint64_t totalRec, balance, donateCnt;
        std::string totalRecStr, balanceStr, donateCntStr;
        if (!ctx->get_object(DONATECOUNT, &donateCntStr) ||
            !ctx->get_object(TOTALRECEIVED, &totalRecStr) ||
            !ctx->get_object(BALANCE, &balanceStr)) {
            ctx->error("read history failed");
            return;
        }
        safe_stoull(totalRecStr, &totalRec);
        safe_stoull(balanceStr, &balance);
        safe_stoull(donateCntStr, &donateCnt);
        totalRec += amount;
        balance += amount;
        donateCnt++;

        totalRecStr = std::to_string(totalRec);
        balanceStr = std::to_string(balance);
        donateCntStr = std::to_string(donateCnt);

        std::string donateID = getIDFromNum(donateCnt);

        std::string userDonateKey = USERDONATE + donor + "/" + donateID;
        std::string allDonateKey = ALLDONATE + donateID;
        std::string donateDetail =
            "donor=" + donor + "," + "amount=" + amountStr + ", " +
            "timestamp=" + timestamp + ", " + "comments=" + comments;
        if (!ctx->put_object(userDonateKey, donateDetail) ||
            !ctx->put_object(allDonateKey, donateDetail) ||
            !ctx->put_object(DONATECOUNT, donateCntStr) ||
            !ctx->put_object(TOTALRECEIVED, totalRecStr) ||
            !ctx->put_object(BALANCE, balanceStr)) {
            ctx->error("failed to save donate record");
            return;
        }
        ctx->ok(donateID);
    }

    void cost() {
        xchain::Context* ctx = this->context();
        const std::string& caller = ctx->initiator();
        if (caller.empty()) {
            ctx->error("missing initiator");
            return;
        }

        if (!isAdmin(ctx, caller)) {
            ctx->error("only the admin can add cost record");
            return;
        }

        const std::string& to = ctx->arg("to");
        if (to.empty()) {
            ctx->error("missing 'to'");
            return;
        }

        const std::string& amountStr = ctx->arg("amount");
        if (amountStr.empty()) {
            ctx->error("missing 'amount'");
            return;
        }

        uint64_t amount;
        if (!safe_stoull(amountStr, &amount)) {
            ctx->error(
                "illegel 'amount', should be string of a positive number");
            return;
        }

        const std::string& timestamp = ctx->arg("timestamp");
        if (timestamp.empty()) {
            ctx->error("missing 'timestamp'");
            return;
        }

        const std::string& comments = ctx->arg("comments");
        if (comments.empty()) {
            ctx->error("missing 'comments'");
            return;
        }

        uint64_t totalCost, balance, costCnt;
        std::string totalCostStr, balanceStr, costCntStr;
        if (!ctx->get_object(COSTCOUNT, &costCntStr) ||
            !ctx->get_object(TOTALCOSTS, &totalCostStr) ||
            !ctx->get_object(BALANCE, &balanceStr)) {
            ctx->error("read history failed");
            return;
        }
        safe_stoull(totalCostStr, &totalCost);
        safe_stoull(balanceStr, &balance);
        safe_stoull(costCntStr, &costCnt);
        if (balance < amount) {
            ctx->error("fund balance is not enough");
            return;
        }

        totalCost += amount;
        balance -= amount;
        costCnt++;

        totalCostStr = std::to_string(totalCost);
        balanceStr = std::to_string(balance);
        costCntStr = std::to_string(costCnt);

        std::string costID = getIDFromNum(costCnt);

        std::string allCostKey = ALLCOST + costID;
        std::string costDetail = "to=" + to + "," + "amount=" + amountStr +
                                 ", " + "timestamp=" + timestamp + ", " +
                                 "comments=" + comments;
        if (!ctx->put_object(allCostKey, costDetail) ||
            !ctx->put_object(COSTCOUNT, costCntStr) ||
            !ctx->put_object(TOTALCOSTS, totalCostStr) ||
            !ctx->put_object(BALANCE, balanceStr)) {
            ctx->error("failed to save cost record");
            return;
        }
        ctx->ok(costID);
    }

    void statistics() {
        xchain::Context* ctx = this->context();
        uint64_t totalCost, balance, totalRec;
        std::string totalCostStr, balanceStr, totalRecStr;
        if (!ctx->get_object(TOTALRECEIVED, &totalRecStr) ||
            !ctx->get_object(TOTALCOSTS, &totalCostStr) ||
            !ctx->get_object(BALANCE, &balanceStr)) {
            ctx->error("read history failed");
            return;
        }

        std::string result = "totalDonates=" + totalRecStr + ", " +
                             "totalCosts=" + totalCostStr + ", " +
                             "fundBalance=" + balanceStr + "\n";
        ctx->ok(result);
    }

    void queryDonor() {
        xchain::Context* ctx = this->context();
        // admin can get the asset data of other users
        const std::string& donor = ctx->arg("donor");
        if (donor.empty()) {
            ctx->error("missing 'donor' in request");
            return;
        }

        if (donor.find("/") != std::string::npos) {
            ctx->error("donor should not contain /");
            return;
        }

        std::string userDonateKey = USERDONATE + donor + "/";
        std::unique_ptr<xchain::Iterator> iter =
            ctx->new_iterator(userDonateKey, userDonateKey + "~");
        std::string result;
        int donateCnt = 0;
        while (iter->next()) {
            std::pair<std::string, std::string> res;
            iter->get(&res);
            if (res.first.length() > userDonateKey.length()) {
                donateCnt++;
                std::string donateID = res.first.substr(userDonateKey.length());
                std::string content = res.second;
                result += "id=" + donateID + ", " + content + "\n";
            }
        }
        result =
            "total donate count:" + std::to_string(donateCnt) + "\n" + result;
        ctx->ok(result);
    }

    void queryDonates() {
        xchain::Context* ctx = this->context();
        const std::string& startID = ctx->arg("startid");
        if (startID.empty()) {
            ctx->error("missing 'startid' in request");
            return;
        }

        const std::string& limitStr = ctx->arg("limit");
        if (limitStr.empty()) {
            ctx->error("missing 'limit' in request");
            return;
        }
        uint64_t limit;
        safe_stoull(limitStr, &limit);
        if (limit > MAX_LIMIT) {
            ctx->error("'limit' is too large");
            return;
        }

        std::string donateKey = ALLDONATE + startID;
        std::string result;
        std::unique_ptr<xchain::Iterator> iter =
            ctx->new_iterator(donateKey, ALLDONATE + "~");
        int selected = 0;
        while (iter->next() && selected < limit) {
            std::pair<std::string, std::string> res;
            iter->get(&res);
            if (res.first.length() > ALLDONATE.length()) {
                selected++;
                std::string donateID = res.first.substr(ALLDONATE.length());
                std::string content = res.second;
                result += "id=" + donateID + ", " + content + '\n';
            }
        }
        ctx->ok(result);
    }

    void queryCosts() {
        xchain::Context* ctx = this->context();
        const std::string& startID = ctx->arg("startid");
        if (startID.empty()) {
            ctx->error("missing 'startid' in request");
            return;
        }

        const std::string& limitStr = ctx->arg("limit");
        if (limitStr.empty()) {
            ctx->error("missing 'limit' in request");
            return;
        }
        uint64_t limit;
        safe_stoull(limitStr, &limit);
        if (limit > MAX_LIMIT) {
            ctx->error("'limit' is too large");
            return;
        }

        std::string costKey = ALLCOST + startID;
        std::string result;
        std::unique_ptr<xchain::Iterator> iter =
            ctx->new_iterator(costKey, ALLCOST + "~");
        int selected = 0;
        while (iter->next() && selected < limit) {
            std::pair<std::string, std::string> res;
            iter->get(&res);
            if (res.first.length() > ALLCOST.length()) {
                selected++;
                std::string costID = res.first.substr(ALLCOST.length());
                std::string content = res.second;
                result += "id=" + costID + ", " + content + '\n';
            }
        }
        ctx->ok(result);
    };
};

DEFINE_METHOD(CharityDemo, initialize) { self.initialize(); }

DEFINE_METHOD(CharityDemo, donate) { self.donate(); }

DEFINE_METHOD(CharityDemo, cost) { self.cost(); }

DEFINE_METHOD(CharityDemo, statistics) { self.statistics(); }

DEFINE_METHOD(CharityDemo, queryDonor) { self.queryDonor(); }

DEFINE_METHOD(CharityDemo, queryDonates) { self.queryDonates(); }

DEFINE_METHOD(CharityDemo, queryCosts) { self.queryCosts(); }